NameToken example

annenkov opened the pull request:
## Purpose

Adds a new NameToken example. NameToken is based on an NFT token example extended with additional methods for managing names.
Closes #169 

## Changes

- `examples/nametoken` folder contains the example.


## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

DOBEN commented on examples/nametoken/src/lib.rs:
```diff
@@ -0,0 +1,1719 @@
+    // expiration date
+    name_expires: Timestamp,
+    // associated data `StateBox` allows for lazy loading data; this is helpful
+    // in the sutuations when one wants to do a partial update not touching
```
`situations` instead of `sutuations`

DOBEN commented on examples/nametoken/src/lib.rs:
```diff
@@ -0,0 +1,1719 @@
+// Data update fee in CCD
+const UPDATE_FEE: Amount = Amount::from_ccd(7);
+
+// Extension fee in CCD
```
Suggestion:
The comment could use “renewal fee” instead of "extension fee" to be consistent.

DOBEN commented on examples/nametoken/src/lib.rs:
```diff
@@ -0,0 +1,1719 @@
+    OwnerShouldBeAccount,
+    /// The amount does not match the fee
+    IncorrectFee,
+    // some invarians of the state are broken
```
Please use `///` instead of `//`. Can you also please use a capital letter for the first letter to be consistent.

annenkov commented:
I think I forgot one thing: registration/renewal/update requires a fee, but nobody can actually withdraw the money :) I guess the contract instance should have an owner, who can do that.

annenkov commented:
In fact, any contract already has an owner, I guess I could just add an entrypoint to transfer money to the owner's account.

DOBEN commented:
The contract already has an owner (which is the account that initialized the contract) but this account can not be changed anymore. 
Suggestion: You can introduce a new variable ‘owner/admin’. this ‘owner/admin` can withdraw the fees and this ‘owner/admin can `transfer` the ownership to a new address (adding a function to transfer the ownership). This setup has the advantage that e.g. a developer can deploy the contracts and then the ownership of the contracts can be passed on over time depending on the situations that arise (e.g. employer leaves the company and the next person in charge takes over).

DOBEN commented:
Tip: You can write ‘closes #169` in the description of this merge request. This will link your issue to this MR and it will also automatically close the issue once this MR is merged.

annenkov commented on examples/nametoken/src/lib.rs:
```diff
@@ -0,0 +1,1848 @@
+
+/// View function that returns the entire contents of the state. Meant for
+/// testing.
+#[receive(contract = "nametoken", name = "view", return_value = "ViewState")]
```
Do we have a convention on how to name contracts? I also don't have a strong opinion. Words in names in `Cargo.toml` are mostly separated with a dash, like "cis2-nft", but it seems there is no consistency between this name and the one in the code, `contract = "CIS2-NFT"`. Other examples use different conventions.
The documentation uses "NameToken" (starting with a capital). I can then use this name consistently  both in `Cargo.toml` and in receive/init attrubutes.

annenkov commented on examples/nametoken/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+default = ["std", "crypto-primitives"]
+crypto-primitives = ["concordium-std/crypto-primitives"]
+std = ["concordium-std/std", "concordium-cis2/std"]
+
```
I use `cargo-concordium 2.1.0`

annenkov replied:
@DOBEN Is it resolved now?

DOBEN replied:
yes, it is working now. I had to install some missing package on my side.

annenkov commented on examples/nametoken/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "nametoken"
```
I use `NameToken` consistently now

limemloh commented on examples/nametoken/src/lib.rs:
```diff
@@ -0,0 +1,1849 @@
+    ) -> ContractResult<()> {
+        ensure!(self.contains_token(token_id), ContractError::InvalidTokenId);
+        // A zero transfer does not modify the state.
+        if amount == 0.into() {
+            return Ok(());
+        }
```
You should check the `from` and `to` address before this point, since a zero transfer between contracts would succeed right now.
I know it is nitpicking, but CIS-2 have a requirement that zero transfers must be treated as a normal transfer (seen from outside of the contract)

annenkov replied:
Yes, fair enough; fixed.

annenkov commented on examples/nametoken/src/lib.rs:
```diff
@@ -0,0 +1,1849 @@
+const RENEWAL_FEE: Amount = Amount::from_ccd(7);
+
+// How long the registered name is owned before it needs to be renewed
+const REGISTRATION_PERIOD_DAYS: u64 = 365;
```
@limemloh do you mean to make this part of the state and let the users initialise it in `init`?

limemloh replied:
No, I mean changing this line to something like:
```suggestion
const REGISTRATION_PERIOD_DAYS: Duration = Duration::from_days(365);
```
But this will currently result in a compile error, since we did not mark `Duration::from_days` as a [`const`-function](https://doc.rust-lang.org/std/keyword.const.html#compile-time-evaluable-functions).
So you would have to make a another PR adding the `const` mark before doing this

annenkov replied:
Ah, ok. I'll leave it as it is for now here and make a PR to `concordium-contracts-common`. We can update the example later on.

annenkov commented on examples/nametoken/src/lib.rs:
```diff
@@ -0,0 +1,1849 @@
+            .ok_or(ContractError::Custom(CustomContractError::InvokeContractError))?;
+        // update expiration date and replace old data with an empty vector
+        self.all_names
+            .get_mut(name)
```
Ok, I've tried to just look up once and get a mutable reference, but now the compiler complains that `self.transfer` is trying to borrow `*self` as mutable, but it is borrowed as immutable when I look up for the first time
```rust
ln register_expired(
        &mut self,
        now: Timestamp,
        name: ContractTokenId,
        owner: AccountAddress,
        state_builder: &mut StateBuilder<S>,
    ) -> ContractResult<AccountAddress> {
        let mut name_info = self
            .all_names
            .get_mut(&name)
            .ok_or(ContractError::Custom(CustomContractError::InconsistentState))?;
        let old_expires = name_info.name_expires;
        // check whether the name has expired
        ensure!(now > old_expires, CustomContractError::NameIsTaken.into());

        let old_owner = name_info.owner;

        // transfer ownership
       // cannot borrow `*self` as mutable because it is also borrowed as immutable
        self.transfer(
            &name,
            1.into(),
            &Address::Account(old_owner),
            &Address::Account(owner),
            state_builder,
        )?;
        let new_expires = now
            .checked_add(Duration::from_days(REGISTRATION_PERIOD_DAYS))
            .ok_or(ContractError::Custom(CustomContractError::OverflowError))?;
        // update expiration date and replace old data with an empty vector
        name_info.name_expires = new_expires;
        name_info.data.replace(Vec::new());
        Ok(old_owner)
}
```

annenkov replied:
No, I still run into the same problem.

annenkov replied:
It would work, if the `self.transfer` call is the last statement. But in this case, I cannot return the old owner.

abizjak replied:
If you add a `        drop(name_info)` before the self.transfer that should work.

abizjak replied:
Also you should probably use `data.clear()` even though it is not technically the same.

annenkov replied:
So, `data.clear()` removes all elements in the vector. But I use `StateBox`, would calling `clear()` require first loading the data? Is there a way of resetting the data to an empty vector without loading the previous content?

annenkov replied:
`drop` fixed the issue, indeed.
For `data` I ended up with replacing it with a new vector and deleting old data. Potentially, it could be cheaper than loading the vector, but I haven't measured :)

annenkov commented:
I've addressed the comments.

One question remains: once a name is transferred, the old data is not erased. It is only erased when taking over an expired name. That means that if a name was traded as an NFT, it will be transferred (by the auction contract?) to a new owner with old data.
I'm not sure whether `transfer` should erase data, it's always possible to do this with an extra `update_data` call, but this one requires paying a fee.

annenkov commented:
Ended up just commenting that the associated data is untouched by `transfer` and another choice is to erase it after transfer.

annenkov commented:
Ended up commenting that `transfer` leaves the associated data untouched and another choice is to erase it.
